rdf-delta-server-local/src/main/java/org/seaborne/delta/zk/ExponentialBackoffZkConnection.java

```diff
@@ -0,0 +1,156 @@
+package org.seaborne.delta.zk;
+
+import org.apache.jena.atlas.json.JsonObject;
+import org.apache.zookeeper.CreateMode;
+import org.apache.zookeeper.KeeperException;
+import org.apache.zookeeper.Watcher;
+
+import java.util.List;
+import java.util.function.Supplier;
+
+public final class ExponentialBackoffZkConnection implements ZkConnection {
+    private final ZkConnection connection;
+    private final int retries;
+
+    public ExponentialBackoffZkConnection(final ZkConnection connection, final int retries) {
+        this.connection = connection;
+        this.retries = retries;
+    }
+
+    @FunctionalInterface
+    private interface ZkMethod<T> {
+        T evaluate() throws Exception;
+    }
+
+    @FunctionalInterface
+    private interface ZkProc {
+        void evaluate() throws Exception;
+    }
+
+    private <T> T retry(final ZkMethod<T> method) throws Exception {
+        T result = null;
+        var i = 0;
+        while (result == null) {
+            try {
+                result = method.evaluate();
+            } catch (final KeeperException e) {
+                if (i == this.retries) {
+                    throw e;
+                } else {
+                    Thread.sleep((long) Math.pow(2, i));
+                }
+            }
+            i++;
+        }
+        return result;
+    }
+
+    private void retry(final ZkProc method) throws Exception {
+        retry(() -> { method.evaluate(); return 0; });
+    }
+
+    private void uncheckedRetry(final Runnable method) {
+        uncheckedRetry(() -> { method.run(); return 0; });
+    }
+
+    private <T> T uncheckedRetry(final Supplier<T> method) {
+        try {
+            return retry(method::get);
+        } catch(final RuntimeException e) {
+            throw e;
+        } catch (final Exception e) {
+            // None expected, however, if one is received:
+            throw new RuntimeException(e);
```

Comment:
It doesn't make sense to create a specific exception here because this code branch is not expected to ever be reached. This code block only exists because of the checked exceptions in the underlying method declarations forcing us to include to satisfy the compiler.